src/java/htsjdk/samtools/SRAIterator.java

```diff
@@ -0,0 +1,195 @@
+*
+* ===========================================================================
+*
+*/
+
+/**
+ * Created by andrii.nikitiuk on 8/11/15.
+ */
+
+package htsjdk.samtools;
+
+import htsjdk.samtools.SAMFileHeader.SortOrder;
+
+import htsjdk.samtools.sra.SRAAlignmentIterator;
+import htsjdk.samtools.sra.SRAUnalignmentIterator;
+import ngs.ErrorMsg;
+import ngs.ReadCollection;
+import ngs.Reference;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class SRAIterator implements SAMRecordIterator {
+    private ValidationStringency validationStringency;
+
+    private ReadCollection run;
+    private SAMFileHeader header;
+    private final List<Reference> cachedReferences;
+    private RecordRangleInfo recordRangleInfo;
+    private Iterator<Chunk> chunksIterator;
+    private Chunk currentChunk;
+
+    private SRAAlignmentIterator alignmentIterator;
+    private SRAUnalignmentIterator unalignmentIterator;
+
+    public static class RecordRangleInfo {
+        private List<Long> referenceOffsets;
+        private List<Long> referenceLengths;
+        private long totalReferencesLength;
+        private long numberOfReads; // is used for unaligned read space
+        private long totalRecordRangeLength;
+
+        public RecordRangleInfo(List<Long> referenceLengths, long numberOfReads) {
+            this.numberOfReads = numberOfReads;
+            this.referenceLengths = referenceLengths;
+
+            referenceOffsets = new ArrayList<Long>();
+
+            totalReferencesLength = 0;
+            for (Long refLen : referenceLengths) {
+                referenceOffsets.add(totalReferencesLength);
+                totalReferencesLength += refLen;
+            }
+
+            totalRecordRangeLength = totalReferencesLength + this.numberOfReads;
+        }
+
+        public long getNumberOfReads() {
+            return numberOfReads;
+        }
+
+        public long getTotalReferencesLength() {
+            return totalReferencesLength;
+        }
+
+        public long getTotalRecordRangeLength() {
+            return totalRecordRangeLength;
+        }
+
+        public final List<Long> getReferenceOffsets() {
+            return referenceOffsets;
+        }
+
+        public final List<Long> getReferenceLengths() {
+            return referenceLengths;
+        }
+    }
+
+    public static RecordRangleInfo getRecordsRangeInfo(ReadCollection run) {
+        try {
+            return new RecordRangleInfo(SRAAlignmentIterator.getReferencesLengths(run), SRAUnalignmentIterator.getNumberOfReads(run));
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public SRAIterator(final ReadCollection run, final SAMFileHeader header, final List<Reference> cachedReferences,
+                       final RecordRangleInfo recordRangleInfo, final List<Chunk> chunks) {
```

Comment:
Reference objects can be easily shared between threads once created. So, it makes sense to cache them within a single SRAFileReader